x/blob/ante/max_blob_size_ante.go

```diff
@@ -0,0 +1,93 @@
+package ante
+
+import (
+	"github.com/celestiaorg/celestia-app/pkg/appconsts"
+	blobtypes "github.com/celestiaorg/celestia-app/x/blob/types"
+
+	"cosmossdk.io/errors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+// MaxBlobSizeDecorator helps to prevent a PFB from being included in a block
+// but not fitting in a data square.
+type MaxBlobSizeDecorator struct {
+	k BlobKeeper
+}
+
+func NewMaxBlobSizeDecorator(k BlobKeeper) MaxBlobSizeDecorator {
+	return MaxBlobSizeDecorator{k}
+}
+
+// AnteHandle implements the AnteHandler interface. It returns an error if the
+// tx contains a MsgPayForBlobs where the total blob data size exceeds the upper
+// bound. The upper bound is calculated based on the number of bytes in a data
+// square with the maximum square size.
+func (d MaxBlobSizeDecorator) AnteHandle(ctx sdk.Context, tx sdk.Tx, simulate bool, next sdk.AnteHandler) (sdk.Context, error) {
+	if ctx.IsReCheckTx() {
```

Comment:
[discussion]
imo we shouldn't run this during deliver tx, since 1) this would technically be adding block validity rule since it runs on PFBs 2) the only way a PFB can make it in the block is if the blob can fit. If a PFB gets included, but the blob doesn't, then we have to rely on block validity rules (via constructing the square) since by DeliverTx the PFB will have already been included and a blob inclusion fraud proof will be created.

this is mostly an efficiency thing, but I don't see any benefit atm, wdyt?
```suggestion
	if !ctx.IsCheckTx() {
```

Reply:
we could also be more specific here, like `if ctx.IsReCheckTx || if ctx.IsDeliverTx` if we don't like `if !ctx.CheckTx`

Reply:
Hmm, I agree it doesn't seem necessary to run this ante handler on `DeliverTx()` b/c that is adding a block validity rule. However, I wonder if it makes sense to run this ante handler on `ReCheckTx()`. Consider the following scenario:

Assume that gov max square size is 128 at t=0.
t=0: User submits txA that is a large PFB. txA passes ante handler and gets gossiped to network. txA is valid but doesn't get included in a block yet.
t=1: New block lands that decreases gov max square size from 128 to 64
t=2: txA remains in mempool. Validators can't include it b/c it can only fit in a square size of 128.

I assume that txA will eventually be evicted from the mempool. The situation is a bit contrived and I think it probably makes more sense to only run this ante handler once on `IsCheckTx()`.


Note: our version of ctx doesn't have `ctx.IsDeliverTx()` 

<img width="964" alt="Screenshot 2023-08-07 at 2 35 19 PM" src="https://github.com/celestiaorg/celestia-app/assets/3699047/5d4a0b88-2759-4aa0-a9d4-3f10d609a928">

so even though I prefer the more explicit:

```go
	if ctx.IsReCheckTx() || ctx.IsDeliverTx()
```

I'll go with

```go
	if !ctx.IsCheckTx() {
```

Reply:
Resolved by https://github.com/celestiaorg/celestia-app/pull/2202/commits/f8914cb53907c0023374be2cf050ec14b6b2d7e5